subgraphs/compound-v3/src/sdk/manager.ts

```diff
@@ -0,0 +1,902 @@
+    this.snapshots.updateUsageData(TransactionType.FLASHLOAN, account);
+
+    return flashloan;
+  }
+
+  //////////////////
+  //// Updaters ////
+  //////////////////
+
+  // used to update tvl, borrow balance, reserves, etc. in market and protocol
+  updateMarketAndProtocolData(
+    inputTokenPriceUSD: BigDecimal,
+    newInputTokenBalance: BigInt,
+    newVariableBorrowBalance: BigInt | null = null,
+    newStableBorrowBalance: BigInt | null = null,
+    newReserveBalance: BigInt | null = null,
+    exchangeRate: BigDecimal | null = null
+  ): void {
+    const mantissaFactorBD = exponentToBigDecimal(
+      this.inputToken.getDecimals()
+    );
+    this.inputToken.updatePrice(inputTokenPriceUSD);
+    this.market.inputTokenPriceUSD = inputTokenPriceUSD;
+    this.market.inputTokenBalance = newInputTokenBalance;
+    if (newVariableBorrowBalance) {
+      this.market.variableBorrowedTokenBalance = newVariableBorrowBalance;
+    }
+    if (newStableBorrowBalance) {
+      this.market.stableBorrowedTokenBalance = newStableBorrowBalance;
+    }
+    if (newReserveBalance) {
+      this.market.reserves = newReserveBalance
+        .toBigDecimal()
+        .div(mantissaFactorBD)
+        .times(inputTokenPriceUSD);
+    }
+    if (exchangeRate) {
+      this.market.exchangeRate = exchangeRate;
+    }
+    const vBorrowAmount = this.market.variableBorrowedTokenBalance
+      ? this.market
+          .variableBorrowedTokenBalance!.toBigDecimal()
+          .div(mantissaFactorBD)
+      : BIGDECIMAL_ZERO;
+    const sBorrowAmount = this.market.stableBorrowedTokenBalance
+      ? this.market
+          .stableBorrowedTokenBalance!.toBigDecimal()
+          .div(mantissaFactorBD)
+      : BIGDECIMAL_ZERO;
+    const totalBorrowed = vBorrowAmount.plus(sBorrowAmount);
+    this.market.totalValueLockedUSD = newInputTokenBalance
+      .toBigDecimal()
+      .div(mantissaFactorBD)
+      .times(inputTokenPriceUSD);
+    this.market.totalBorrowBalanceUSD = totalBorrowed.times(inputTokenPriceUSD);
+    this.market.save();
+
+    let totalValueLockedUSD = BIGDECIMAL_ZERO;
+    let totalBorrowBalanceUSD = BIGDECIMAL_ZERO;
+    const marketList = this.protocol._markets;
+    for (let i = 0; i < marketList.length; i++) {
+      const _market = Market.load(marketList[i]);
+      if (!_market) {
+        log.error("[updateMarketAndProtocolData] Market not found: {}", [
+          marketList[i].toHexString(),
+        ]);
+        continue;
+      }
+      totalValueLockedUSD = totalValueLockedUSD.plus(
+        _market.totalValueLockedUSD
+      );
+      totalBorrowBalanceUSD = totalBorrowBalanceUSD.plus(
+        _market.totalBorrowBalanceUSD
+      );
+    }
+    this.protocol.totalValueLockedUSD = totalValueLockedUSD;
+    this.protocol.totalBorrowBalanceUSD = totalBorrowBalanceUSD;
+    this.protocol.save();
+  }
+
+  updateRevenue(
+    newTotalRevenueUSD: BigDecimal,
+    newProtocolRevenueUSD: BigDecimal,
+    newSupplySideRevenueUSD: BigDecimal,
```

Comment:
Is `total` always `protocol + supply`? If so you can omit the total.

Also, from the naming it looks as if these revenues are going to replace the old ones. Maybe rename to `addRevenue`, or the arguments to revenue deltas?

Reply:
this is true, fixed